bin/runtime-common/src/messages_xcm_extension.rs

```diff
@@ -144,155 +141,199 @@ impl SenderAndLane {
+			sender_and_lane.location,
+			sender_and_lane.lane,
+			enqueued_messages,
+		);
+
+		if let Err(e) = Self::send_congested_signal(sender_and_lane) {
+			log::info!(
+				target: crate::LOG_TARGET_BRIDGE_DISPATCH,
+				"Failed to send the 'congested' XCM message to {:?}: {:?}",
+				sender_and_lane.location,
+				e,
+			);
+		}
 	}
-}
 
-/// A structure that implements [`frame_support::traits::QueuePausedQuery`] and may
-/// be used in the `pallet-message-queue` configuration to stop processing messages when the
-/// bridge queue is congested.
-///
-/// It needs to be used at the source bridge hub.
-pub struct LocalXcmQueueSuspender<Origin, Inner, R, MI, SL>(
-	PhantomData<(Origin, Inner, R, MI, SL)>,
-);
+	/// Must be called whenever we receive a message delivery confirmation.
+	pub fn on_bridge_messages_delivered(
+		sender_and_lane: &SenderAndLane,
+		enqueued_messages: MessageNonce,
+	) {
+		// if we have not sent the congestion signal before, we don't want to do anything
+		if !Self::is_congested_signal_sent(sender_and_lane.lane) {
+			return
+		}
 
-impl<Origin, Inner, R, MI, SL> QueuePausedQuery<Origin>
-	for LocalXcmQueueSuspender<Origin, Inner, R, MI, SL>
-where
-	Origin: Clone + Into<MultiLocation>,
-	Inner: QueuePausedQuery<Origin>,
-	R: MessagesConfig<MI>,
-	MI: 'static,
-	SL: Get<SenderAndLane>,
-{
-	fn is_paused(origin: &Origin) -> bool {
-		// give priority to inner status
-		if Inner::is_paused(origin) {
-			return true
+		// if the bridge queue is still congested, we don't want to do anything
+		let is_congested = enqueued_messages > OUTBOUND_LANE_UNCONGESTED_THRESHOLD;
+		if is_congested {
+			return
 		}
 
-		// if we have suspended the queue before, do not even start processing its messages
-		let sender_and_lane = SL::get();
-		let is_expected_origin = origin.clone().into() == sender_and_lane.location;
-		if is_expected_origin &&
-			LocalXcmQueueManager::is_inbound_queue_suspended::<R, MI>(sender_and_lane.lane)
-		{
-			return true
+		log::info!(
+			target: crate::LOG_TARGET_BRIDGE_DISPATCH,
+			"Sending 'uncongested' XCM message to {:?}. Lane {:?}: there are\
+			{} messages queued at the bridge queue",
+			sender_and_lane.location,
+			sender_and_lane.lane,
+			enqueued_messages,
+		);
+
+		if let Err(e) = Self::send_uncongested_signal(sender_and_lane) {
+			log::info!(
+				target: crate::LOG_TARGET_BRIDGE_DISPATCH,
+				"Failed to send the 'uncongested' XCM message to {:?}: {:?}",
+				sender_and_lane.location,
+				e,
+			);
 		}
+	}
+
+	/// Returns true if we have sent "congested" signal to the `sending_chain_location`.
+	fn is_congested_signal_sent(lane: LaneId) -> bool {
+		OutboundLanesCongestedSignals::<H::Runtime, H::MessagesInstance>::get(lane)
+	}
+
+	/// Send congested signal to the `sending_chain_location`.
+	fn send_congested_signal(sender_and_lane: &SenderAndLane) -> Result<(), SendError> {
+		send_xcm::<H::ToSourceChainSender>(sender_and_lane.location, H::CongestedMessage::get())?;
```

Comment:
@svyatonik 
last thin, 
I am not sure, if we should better parametrize with `BridgeId/LaneId` now instead of `Get<Xcm<()>>`,
probably for v1 is ok like this

Reply:
(or even better - we'll introduce a logical HRMP subchannels and remove this message sending at all)